### PR TITLE
API Replace old assets/.htaccess with better default

### PR DESCRIPTION
**Pull request by tractorcow**

### CODE REVIEW

**Comment by tractorcow** [comment]
@dhensby this file is regenerated on flush, but the old default file wasn't compatible with the new asset abstraction stuff.

Does it make sense to you? :P Not sure if you check installer issues often heh. :D

File: assets/.htaccess [hunk 1]
```diff
@@ -1,36 +1,27 @@
...
+    SetEnv HTTP_MOD_REWRITE On
+    RewriteEngine On
+
+    # Disable PHP handler
+    RewriteCond %{REQUEST_URI} .(?i:php|phtml|php3|php4|php5|inc)$
+    RewriteRule .* - [F]
+
+    # Allow error pages
+    RewriteCond %{REQUEST_FILENAME} -f
+    RewriteRule error[^\/]*.html$ - [L]
```
Comment (dhensby) [inline assets/.htaccess]: is this going to be error pages ONLY in the root of the assets dir?

Comment (tractorcow) [inline assets/.htaccess]: No, this just means error pages skip any of the below rules and don't need to respect any file extension rules.

Comment (dhensby) [inline assets/.htaccess]: My concern, which may be completely misfounded, is that a malicious user could upload an error-*.html file into the assets and it'd "work" and be presented as some kind of phishing or XSS flaw.

in fact, do we even need these files to be viewable directly by URL!? Aren't they just used as a static cache?

**Comment by tractorcow** [comment]
> My concern, which may be completely misfounded, is that a malicious user could upload an error-*.html file into the assets and it'd "work" and be presented as some kind of phishing or XSS flaw.

Yep that's right; It's completely possible.

> in fact, do we even need these files to be viewable directly by URL!? Aren't they just used as a static cache?

Apache needs to internally rewrite them, so I'm not sure if there's another way we could configure them...

**Comment by dhensby** [comment]
If we re-write to them, do they need to be accessible to the outside? I think we might need to think of a better way to handle error pages if we expose a security hole like this. 

@jonom, weren't you doing some work on error pages?

**Comment by tractorcow** [comment]
I should point out that at the moment, the current assets/.htaccess on the master branch breaks framework; It needs to be manually deleted or flushed to fix. Until we merge this checkouts of 4 are going to continue breaking...

Can we leave the error pages public for the time being and fix this later?

**Comment by dhensby** [comment]
If we're allowing any document in assets named error-*.html, it's a security risk. if we can limit it to being just assets/error-*.html then  that'll be an ok stop gap to patch later

**Comment by tractorcow** [comment]
How about any document `error([-_\w])*\.html`? That'll allow translatable / core files at least.

**Comment by dhensby** [comment]
only in the root of assets? sure.

**Comment by jonom** [comment]
> @jonom, weren't you doing some work on error pages?

Sorry @dhensby I totally missed this mention. Yeah if I get my way there won't be any static error pages in the assets directory since they're next to (or worse than) useless for most site at the moment. [Haven't quite convinced @sminnee of this fact yet](https://github.com/silverstripe/silverstripe-framework/issues/4149#issuecomment-194579085) and gotten the green accepted label but I haven't given up hope ;)

Would still need a static 500 error page somewhere but I would probably stick it in root instead of the assets directory.

File: assets/.htaccess (hunk 1 above)
Comment (dhensby) [inline assets/.htaccess]: Does the browser need direct access to these files? If we block access but php / Apache access them internally to serve on 404/500, will that cause a problem?

Comment (jonom) [inline assets/.htaccess]: I would have thought you were right @dhensby but I've just tried triggering a static 404 on a site and the file doesn't seem to be used if blocked by .htaccess. Steps to test:

1. Prevent SilverStripe handling request by removing the `<IfModule mod_rewrite.c>` section from website's root .htaccess
2. Go to site.dev/asdas or whatever - should load static 404 (which may be [shockingly broken and out of date](https://github.com/silverstripe/silverstripe-framework/issues/4149))
3. Remove `html|htm|xhtml|` from assets/.htaccess FilesMatch whitelist
4. Go to site.dev/asdas or whatever again - should load generic Apache 404 message.

Comment (jonom) [inline assets/.htaccess]: Just tried this with the updated .htacess file proposed here but removing the `Allow error pages` section and same thing happens. Guess Apache makes a request to itself to fetch the static error page.

Comment (dhensby) [inline assets/.htaccess]: Sigh

Comment (jonom) [inline assets/.htaccess]: Maybe you could adapt this little snippet: https://www.digitalocean.com/community/tutorials/how-to-configure-apache-to-use-custom-error-pages-on-ubuntu-14-04#respond-with-404-when-error-pages-are-directly-requested

Comment (dhensby) [inline assets/.htaccess]: Yes, that might do it...

Comment (tractorcow) [inline assets/.htaccess]: > To implement this behavior, we'll need to add a Files block for each of our custom pages. 

Not going to easily happen;

**Comment by dhensby** [comment]
If this is blocking progress we can merge into master and open an issue tagged for the alpha release to patch this issue

**Comment by tractorcow** [comment]
We'll need to fix this before the alpha 1 for sure.

**Comment by dhensby** [comment]
Shall I merger and open an issue?

**Comment by tractorcow** [comment]
Sure. :D

**Comment by tractorcow** [comment]
Or rather, yes please... I need the old htaccess gone in any case, even if this replacement isn't perfect.

And any new replacement will need changing in the framework template, which is what generates this.

**Pull request merged**